### DISCUSSION TITLE
khadas edge v gived partition table error

### DISCUSSION THREAD

**Discussion by vivekjaiswal90** [JeffyCN/meta-rockchip#97] [General]
Hello Jeff,
I used your tutorial
Branch used: kirkstone
machine configuration : [rockchip-rk3399-khadas-edge-v.conf.txt](https://github.com/JeffyCN/meta-rockchip/files/14697097/rockchip-rk3399-khadas-edge-v.conf.txt)

Changes i did  : meta-rockchip/conf/machine/include/rk3399.inc
changed form 4.19 to 5.10 as "bitbake core-image-minimal" gave error
```
PREFERRED_VERSION_linux-rockchip := "5.10%"
LINUXLIBCVERSION := "5.10-custom%"
```
The build was successfull. I get error when did your setp 

Kernel didnot started, when chekcing the UART log: 
[UART_Error_2.log.txt](https://github.com/JeffyCN/meta-rockchip/files/14697204/UART_Error_2.log.txt)
```
Starting kernel ...
"Synchronous Abort" handler, esr 0x02000000
* Reason:        Exception from an unknown reason
```

Do you know what is the issue ? does your reposiotry meta-rockchip supports khadas edge-v device for kirkstone branch?

Next i took the following steps
I have uploadded in emmc  follwoing images created after the successfull build 
I used  boot loader method 1 to update the emmc & used rkdevtool
https://opensource.rock-chips.com/wiki_Boot_option

https://opensource.rock-chips.com/wiki_Boot_option#:~:text=img%0Arkdeveloptool%20rd-,For%C2%A0with%20miniloader,-rkdeveloptool%20db%20rkxx_loader_vx 
```
rkdeveloptool db loader.bin 
rkdeveloptool gpt rkdeveloptool/parameter_gpt.txt 
rkdeveloptool wl 0x4000 uboot.img
rkdeveloptool wl 0x6000 trust.img  
rkdeveloptool wl 0x8000 boot.img
rkdeveloptool wl 0x40000 rootfs.img
rkdeveloptool rd
```
and also packaged boot.img using the methid mentioned here and then uploaded using rkdevtool
https://opensource.rock-chips.com/wiki_Rockchip_Kernel#Install_Boot.2Fkernel:~:text=Folder%20tree%20for%20armv8(rk3399%2C%20rk3328)%3A
```
Structure :
boot
├── extlinux
│   └── extlinux.conf
├── Image
└── rk3399-evb.dtb

Packaged the ffolder to boot_rk3399.image
: genext2fs -b 32768 -B $((32*1024*1024/32768)) -d boot/ -i 8192 -U boot_rk3399.img

uploaded to emmc:
rkdeveloptool wl 0x8000 boot_rk3399.img
```

Error:
[partitionErrorlog.txt](https://github.com/JeffyCN/meta-rockchip/files/14697094/partitionErrorlog.txt)
```
[    3.148455] blk_update_request: I/O error, dev mmcblk2, sector 1 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    3.149850] blk_update_request: I/O error, dev mmcblk2, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    3.151255] blk_update_request: I/O error, dev mmcblk2, sector 3 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
```

Can you help me how to solve error in either of the cases.

**Comment by JeffyCN** [comment]
maybe similar to https://github.com/JeffyCN/meta-rockchip/issues/87


and you can also try to use your original working kernel(flash the non-yocto generated old boot.img)

**Reply by vivekjaiswal90** [reply]
HI @JeffyCN 
you mean i have to do following 
1. Set ROCKCHIP_KERNEL_COMPRESSED = "0" in  local.conf.
2. After the build (bitbake core-image-minimal) is completed
3. sudo upgrade_tool db <IMAGE PATH>/loader.bin
4. sudo upgrade_tool wl 0 <IMAGE PATH>/<IMAGE NAME>.wic
5. If above step donot works then do 5 & 6
6. sudo upgrade_tool di -u <IMAGE PATH>/uboot.img 
7. sudo upgrade_tool di -b <IMAGE PATH>/boot.img

Is my understanding correct?

**Reply by JeffyCN** [reply]
right, and the step 6 7 would be using your old non-yocto images

**Reply by vivekjaiswal90** [reply]
After doing Step1 to Step4 . the kernel got started.
But it got hanged in between.
Log :  [UARTLog_eMMC_kernel_compress0.txt](https://github.com/JeffyCN/meta-rockchip/files/14722579/UARTLog_eMMC_kernel_compress0.txt)
```
[    3.182109] blk_update_request: I/O error, dev mmcblk2, sector 1 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    3.183939] Buffer I/O error on dev mmcblk2, logical block 0, async page read
[    3.187707] blk_update_request: I/O error, dev mmcblk2, sector 2 op 0x0:(READ) flags 0x0 phys_seg 1 prio class 0
[    3.189421] Buffer I/O error on dev mmcblk2, logical block 0, async page read
[    3.190144]  mmcblk2: unable to read partition table

```

sorry regaridng 6 & 7 i doonot have "old non-yocto images"
==> Where can I get it from for khadas edge V? 

Later i try to uplaod the wic file in SD card and here as well the kernel is started but got stucked.
Log : [UARTLog_SDacrd_kernel_compress0.txt](https://github.com/JeffyCN/meta-rockchip/files/14722125/UARTLog_SDacrd_kernel_compress0.txt)

```
[    4.446591] rockchip-pcie f8000000.pcie: supply vpcie1v8 not found, using dummy regulator
[    4.447538] rockchip-pcie f8000000.pcie: supply vpcie0v9 not found, using dummy regulator
[    4.511919] mmc2: new HS400 Enhanced strobe MMC card at address 0001
[    4.513252] mmcblk2: mmc2:0001 BJTD4R 29.1 GiB 
[    4.513845] mmcblk2boot0: mmc2:0001 BJTD4R partition 1 4.00 MiB
[    4.514549] mmcblk2boot1: mmc2:0001 BJTD4R partition 2 4.00 MiB
[    4.515489] mmcblk2rpmb: mmc2:0001 BJTD4R partition 3 4.00 MiB, chardev (238:0)
[    4.518651] Alternate GPT is invalid, using primary GPT.
[    4.519223]  mmcblk2: p1 p2 p3 p4
[    4.554528] vendor storage:20190527 ret = 0
[   37.618880] random: crng init done
[   64.502168] rcu: INFO: rcu_sched detected stalls on CPUs/tasks:
[   64.502705] rcu:     1-...0: (1 GPs behind) idle=9b2/1/0x4000000000000000 softirq=27/27 fqs=6000 
[   64.503473]  (detected by 2, t=18002 jiffies, g=-1103, q=11)
[   64.503979] Task dump for CPU 1:
[   64.504272] task:kworker/u12:0   state:R  running task     stack:    0 pid:    7 ppid:     2 flags:0x0000000a
[   64.505170] Workqueue: events_unbound deferred_probe_work_func
[   64.505694] Call trace:
```

**Reply by JeffyCN** [reply]
Porting kernel would not be easy and may related to the hardware spec.

The preferred way would be trying to find an existing worked kernel image. Usually there would be a worked firmware when you get the board from the manifactor.

**Reply by JeffyCN** [reply]
and maybe you can try to disable some modules(pcie, hdmi, etc.) in the kernel's dts or defconfig

**Reply by vivekjaiswal90** [reply]
thanks for sharing the toughts.. but I am really sorry to say,  we have not received any firmware from the manufacturer board we ordered
so seems have to fix kernelimage oneself
but there might be some way to fix the kernel image.
I am using followig in my machine config 

`KBUILD_DEFCONFIG = "kedges_defconfig"`

follwinf th si pcie & hdmi configuration, i will try to disable them all and try
[kedges_defconfig.txt](https://github.com/JeffyCN/meta-rockchip/files/14726202/kedges_defconfig.txt)

```
CONFIG_PCIEPORTBUS=y
CONFIG_PCIEASPM_POWERSAVE=y
CONFIG_PCIE_ROCKCHIP_HOST=y
CONFIG_PCIE_DW_ROCKCHIP=y
CONFIG_BCMDHD_PCIE=y
CONFIG_PHY_ROCKCHIP_PCIE=y
CONFIG_PHY_ROCKCHIP_SNPS_PCIE3=y

CONFIG_ROCKCHIP_DW_HDMI=y
CONFIG_ROCKCHIP_INNO_HDMI=y
CONFIG_DRM_DW_HDMI_I2S_AUDIO=y
CONFIG_DRM_DW_HDMI_CEC=y
CONFIG_SND_SOC_ROCKCHIP_HDMI=y
CONFIG_PHY_ROCKCHIP_INNO_HDMI=y
CONFIG_PHY_ROCKCHIP_SAMSUNG_HDPTX_HDMI=y
```

**Reply by vivekjaiswal90** [reply]
or do you think i should download this repository and use this dts
https://github.com/khadas/linux/blob/khadas-edges-5.10.y-release-v1.0.0/arch/arm64/boot/dts/rockchip/rk3399-khadas-edge-v.dts

and the make a chnages in 
```
# build/conf/local.conf
SRC_URI:pn-linux-rockchip = " \
        git://${TOPDIR}/../kernel;protocol=file;usehead=1 \
        file://cgroups.cfg \
"
SRCREV:pn-linux-rockchip = "${AUTOREV}"
KBRANCH = "HEAD"
```

**Reply by JeffyCN** [reply]
right, you should use that kernel source.

and you may ask the kernel related issues there, since they would know the details about your board.

**Reply by JeffyCN** [reply]
and you may also try to find a way to build the boot.img from that kernel directly(something like ARCH=arm64 CROSS_COMPILE=xxx make xxx_defconfig xxx.img), this would make kernel debugging more efficient.

**Reply by vivekjaiswal90** [reply]
Thanks @JeffyCN so much for reply.
I have posted my question in khadas linux & khadas u-boot community

https://github.com/khadas/linux/issues/113
https://github.com/khadas/u-boot/issues/44

But there is no response yet. I need a working kernel configuration for edge v rk3399 to be able to proceed. 

>> (something like ARCH=arm64 CROSS_COMPILE=xxx make xxx_defconfig xxx.img

Also i tried above that you said. (as per the link https://opensource.rock-chips.com/wiki_Rockchip_Kernel#Install_Boot.2Fkernel:~:text=linux%2Dgnueabihf%2D%20%2Dj4-,For%20ARM%20V8,-make%20ARCH%3Darm64)
Steps done :
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- kedges_defconfig ==> Succeed
make ARCH=arm64 CROSS_COMPILE=aarch64-linux-gnu- -j4  ==> Succeed

make rk3399-khadas-edge-v.img ==>  Failed
Error: 
`make: *** No rule to make target 'rk3399-khadas-edge-v.img'.  Stop.`
Log : [kernel_creation log_from_linux_repo.txt](https://github.com/JeffyCN/meta-rockchip/files/14756903/kernel_creation.log_from_linux_repo.txt)

**Reply by JeffyCN** [reply]
you may need to set ARCH and CROSS_COMPILE env
